homeassistant/components/smartthings/__init__.py

```diff
@@ -194,6 +178,45 @@
     return all(await asyncio.gather(*tasks))
 
 
+async def async_remove_entry(
+        hass: HomeAssistantType, entry: ConfigEntry) -> None:
+    """Perform clean-up when entry is being removed."""
+    from pysmartthings import SmartThings
+
+    api = SmartThings(async_get_clientsession(hass),
+                      entry.data[CONF_ACCESS_TOKEN])
+
+    # Remove the installed_app, which if already removed raises a 403 error.
+    installed_app_id = entry.data[CONF_INSTALLED_APP_ID]
+    try:
+        await api.delete_installed_app(installed_app_id)
+    except ClientResponseError as ex:
+        if ex.status == 403:
+            _LOGGER.exception("Installed app %s has already been removed",
+                              installed_app_id)
+        else:
+            raise
+    _LOGGER.debug("Removed installed app %s", installed_app_id)
+
+    # Remove the app if not referenced by other entries, which if already
+    # removed raises a 403 error.
+    app_id = entry.data[CONF_APP_ID]
+    app_count = sum(1 for entry in hass.config_entries.async_entries(DOMAIN)
+                    if entry.data[CONF_APP_ID] == app_id)
+    if app_count <= 1:
```

Comment:
Should we remove the app if there are no apps for the app id?

Reply:
Yes, though the condition where `app_count == 0` should be unusual... like `ConfigEntries.async_remove` was called concurrently for the same entry, we should try to remove the SmartThings app regardless.  I inverted the `if` statement to make this look less confusing.